File: apple/WealthWise/WealthWise/Core/Repositories/AccountRepository.swift
//
//  AccountRepository.swift
//  WealthWise
//
//  Created by WealthWise Team on 2025-11-08.
//  Repository pattern for Account data access and sync
//

import Foundation
import SwiftData

/// Repository for Account management with Firebase sync
/// Provides offline-first access with background synchronization
@MainActor
final class AccountRepository: ObservableObject {
    
    // MARK: - Properties
    
    @Published private(set) var accounts: [Account] = []
    @Published private(set) var isLoading = false
    @Published private(set) var error: Error?
    
    private let modelContext: ModelContext
    private let firebaseService: FirebaseService
    
    // MARK: - Initialization
    
    init(modelContext: ModelContext, firebaseService: FirebaseService = .shared) {
        self.modelContext = modelContext
        self.firebaseService = firebaseService
    }
    
    // MARK: - Fetch Operations
    
    /// Fetch all accounts from local storage
    func fetchLocal() throws {
        let descriptor = FetchDescriptor<Account>(
            sortBy: [SortDescriptor(\.createdAt)]
        )
        accounts = try modelContext.fetch(descriptor)
    }
    
    /// Sync with Firebase and update local storage
    func sync() async throws {
        isLoading = true
        defer { isLoading = false }
        
        do {
            // Fetch from Firebase
            let remoteAccounts = try await firebaseService.fetchAccounts()
            
            // Clear local storage
            try modelContext.delete(model: Account.self)
            
            // Insert remote accounts
            for account in remoteAccounts {
                modelContext.insert(account)
            }
            
            try modelContext.save()
            
            // Update published property
            try fetchLocal()
            
            error = nil
        } catch {
            self.error = error
            throw error
        }
    }
    
    /// Fetch account by ID
    func fetchById(_ id: UUID) -> Account? {
        let descriptor = FetchDescriptor<Account>(
            predicate: #Predicate { $0.id == id }
        )
        return try? modelContext.fetch(descriptor).first
    }
    
    // MARK: - Create/Update Operations
    
    /// Create new account
    func create(_ account: Account) async throws {
        // Save locally
        modelContext.insert(account)
        try modelContext.save()
        
        // Sync to Firebase
        try await firebaseService.saveAccount(account)
        
        // Refresh local list
        try fetchLocal()
    }
    
    /// Update existing account
    func update(_ account: Account) async throws {
        account.updatedAt = Date()
        try modelContext.save()
        
        // Sync to Firebase
        try await firebaseService.saveAccount(account)
        
        // Refresh local list
        try fetchLocal()
    }
    
    // MARK: - Delete Operations
    
    /// Delete account
    func delete(_ account: Account) async throws {
        // Delete from Firebase
        try await firebaseService.deleteAccount(account.id)
        
        // Delete locally
        modelContext.delete(account)
        try modelContext.save()
        
        // Refresh local list
        try fetchLocal()
    }
    
    /// Archive account (soft delete)
    func archive(_ account: Account) async throws {
        account.isArchived = true
        try await update(account)
    }
    
    /// Unarchive account
    func unarchive(_ account: Account) async throws {
        account.isArchived = false
        try await update(account)
    }
    
    // MARK: - Helper Methods
    
    /// Get active (non-archived) accounts
    func activeAccounts() -> [Account] {
        accounts.filter { !$0.isArchived }
    }
    
    /// Get accounts by type
    func accounts(ofType type: Account.AccountType) -> [Account] {
        accounts.filter { $0.type == type }
    }
    
    /// Calculate total balance across all accounts
    func totalBalance() -> Decimal {
        activeAccounts().reduce(Decimal(0)) { $0 + $1.currentBalance }
    }
    
    /// Recalculate balances for all accounts
    func recalculateAllBalances() async throws {
        for account in accounts {
            account.recalculateBalance()
        }
        try modelContext.save()
        
        // Sync updated balances to Firebase
        for account in accounts {
            try await firebaseService.saveAccount(account)
        }
    }
}

// MARK: - Static Accessor

extension AccountRepository {
    static func create(with modelContext: ModelContext) -> AccountRepository {
        AccountRepository(modelContext: modelContext)
    }
}

File: android/app/src/main/kotlin/com/wealthwise/android/data/repository/AccountRepository.kt
package com.wealthwise.android.data.repository

import com.wealthwise.android.data.local.dao.AccountDao
import com.wealthwise.android.data.model.Account
import com.wealthwise.android.data.remote.firebase.FirestoreService
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.first
import java.math.BigDecimal
import java.time.LocalDateTime
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Repository for account data operations.
 * 
 * Implements offline-first pattern:
 * - All reads come from local database
 * - All writes go to local database first, then sync to Firestore
 * - Conflicts resolved using last-write-wins strategy
 */
@Singleton
class AccountRepository @Inject constructor(
    private val accountDao: AccountDao,
    private val firestoreService: FirestoreService
) {
    
    /**
     * Get all accounts as a Flow.
     * Returns data from local database for instant display.
     */
    fun getAllAccounts(): Flow<List<Account>> {
        return accountDao.getAllFlow()
    }
    
    /**
     * Get active (non-archived) accounts.
     */
    fun getActiveAccounts(): Flow<List<Account>> {
        return accountDao.getActiveAccountsFlow()
    }
    
    /**
     * Get a single account by ID.
     */
    suspend fun getAccountById(accountId: String): Account? {
        return accountDao.getById(accountId)
    }
    
    /**
     * Get total balance across all active accounts.
     */
    suspend fun getTotalBalance(): BigDecimal {
        return accountDao.getTotalBalance()
    }
    
    /**
     * Create a new account.
     * 
     * @param userId User ID
     * @param name Account name
     * @param type Account type
     * @param institution Bank/institution name
     * @param initialBalance Initial account balance
     * @param currency Currency code (default: INR)
     * @return Result containing the created account or error
     */
    suspend fun createAccount(
        userId: String,
        name: String,
        type: Account.AccountType,
        institution: String?,
        initialBalance: BigDecimal,
        currency: String = "INR"
    ): Result<Account> {
        return try {
            val account = Account(
                id = UUID.randomUUID().toString(),
                userId = userId,
                name = name,
                type = type,
                institution = institution,
                currentBalance = initialBalance,
                currency = currency,
                isArchived = false,
                createdAt = LocalDateTime.now(),
                updatedAt = LocalDateTime.now(),
                lastSyncedAt = null // Will be synced
            )
            
            // Insert into local database
            accountDao.insert(account)
            
            // Sync to Firestore in background
            syncAccountToFirestore(account)
            
            Result.success(account)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Update an existing account.
     */
    suspend fun updateAccount(account: Account): Result<Unit> {
        return try {
            val updatedAccount = account.copy(
                updatedAt = LocalDateTime.now(),
                lastSyncedAt = null // Mark as needs sync
            )
            
            accountDao.update(updatedAccount)
            syncAccountToFirestore(updatedAccount)
            
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Update account balance.
     */
    suspend fun updateBalance(accountId: String, newBalance: BigDecimal): Result<Unit> {
        return try {
            accountDao.updateBalance(accountId, newBalance, LocalDateTime.now())
            
            // Get updated account and sync
            val account = accountDao.getById(accountId)
            if (account != null) {
                syncAccountToFirestore(account)
            }
            
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Archive an account.
     */
    suspend fun archiveAccount(accountId: String): Result<Unit> {
        return try {
            val account = accountDao.getById(accountId) ?: return Result.failure(
                Exception("Account not found")
            )
            
            val archivedAccount = account.copy(
                isArchived = true,
                updatedAt = LocalDateTime.now(),
                lastSyncedAt = null
            )
            
            accountDao.update(archivedAccount)
            syncAccountToFirestore(archivedAccount)
            
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Unarchive an account.
     */
    suspend fun unarchiveAccount(accountId: String): Result<Unit> {
        return try {
            val account = accountDao.getById(accountId) ?: return Result.failure(
                Exception("Account not found")
            )
            
            val unarchivedAccount = account.copy(
                isArchived = false,
                updatedAt = LocalDateTime.now(),
                lastSyncedAt = null
            )
            
            accountDao.update(unarchivedAccount)
            syncAccountToFirestore(unarchivedAccount)
            
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Delete an account and all its transactions.
     */
    suspend fun deleteAccount(accountId: String): Result<Unit> {
        return try {
            // Delete from local database (cascade delete will handle transactions)
            accountDao.deleteById(accountId)
            
            // Delete from Firestore
            firestoreService.deleteAccount(accountId)
            
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    /**
     * Search accounts by name or institution.
     */
    fun searchAccounts(query: String): Flow<List<Account>> {
        return accountDao.searchAccounts(query)
    }
    
    /**
     * Sync account to Firestore.
     * Runs in background and doesn't block UI.
     */
    private suspend fun syncAccountToFirestore(account: Account) {
        try {
            val result = firestoreService.updateAccount(account)
            if (result.isSuccess) {
                // Mark as synced in local database
                accountDao.markSynced(account.id, LocalDateTime.now())
            }
        } catch (e: Exception) {
            // Silent failure - will retry on next sync
            // TODO: Add to sync queue for retry
        }
    }
    
    /**
     * Force sync all accounts that need syncing.
     */
    suspend fun syncPendingAccounts(): Result<Int> {
        return try {
            val accounts = accountDao.getAccountsNeedingSync()
            var syncedCount = 0
            
            accounts.forEach { account ->
                val result = firestoreService.updateAccount(account)
                if (result.isSuccess) {
                    accountDao.markSynced(account.id, LocalDateTime.now())
                    syncedCount++
                }
            }
            
            Result.success(syncedCount)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}
